File: 0061. Rotate List.js
const rotateRight = (head, k) => {
  if (head === null || head.next === null || k === 0) return head;

  let node = head,
    listLength = 1;

  while (node.next) {
    node = node.next;
    listLength++;
  }

  newTailIdx = listLength - (k % listLength) - 1;

  if (newTailIdx === listLength - 1) return head;

  let newTail = head;
  for (let i = 0; i < newTailIdx; i++) {
    newTail = newTail.next;
  }

  const newHead = newTail.next;
  newTail.next = null;
  node.next = head;

  return newHead;
};

console.log(
  rotateRight(
    {
      val: 1,
      next: {
        val: 2,
        next: { val: 3, next: { val: 4, next: { val: 5, next: null } } },
      },
    },
    2
  )
);

// 0 1 2
// 2 0 1

File: python/0061. Rotate List.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def rotateRight(self, head: ListNode, shift: int) -> ListNode:
        if (head is None):
            return None

        list_length = self.get_list_length(head)
        shift %= list_length

        if shift == 0:
            return head

        curr_node = head

        for _ in range(list_length - shift - 1):
            curr_node = curr_node.next

        new_head = curr_node.next
        curr_node.next = None
        curr_node = new_head

        while curr_node.next:
            curr_node = curr_node.next

        curr_node.next = head

        return new_head

    def get_list_length(self, head: ListNode) -> int:
        list_length = 0

        while head:
            head = head.next
            list_length += 1

        return list_length
